core/src/main/java/fr/sncf/osrd/utils/new_graph/Pathfinding.java

```diff
@@ -0,0 +1,179 @@
+package fr.sncf.osrd.utils.new_graph;
+
+import com.google.common.graph.ImmutableNetwork;
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+import org.jetbrains.annotations.NotNull;
+import java.util.*;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+@SuppressFBWarnings(
+        value = "FE_FLOATING_POINT_EQUALITY",
+        justification = "No arithmetic is done on values where we test for equality, only copies"
+)
+public class Pathfinding<NodeT, EdgeT> {
+
+    /** Pathfinding step */
+    private record Step<EdgeT>(
+            EdgeRange<EdgeT> range, // Next step
+            Step<EdgeT> prev, // Previous step (to construct the result)
+            double totalDistance, // Total distance from the start
+            double weight, // Priority queue weight (could be different from totalDistance to allow for A*)
```

Comment:
can't you just remove the totalDistance field?

Reply:
I plan on switching to an A* relatively soon where this distinction is needed. Having this in mind from the start makes it easier to write

Reply:
as far as I know A* only needs a weight, not the total distance

Reply:
In A*, weigh = distance + heuristic, but you need the distance alone to compute the distance of the next step. I use `step.totalDistance` at several places in the code and those would be wrong if it included the heuristic